Fix bug 1519894: Add Resource menu

mathjazz opened the pull request:
This patch exposes the name of the resource currently localized in the main navigation, and lists the entire batch of project resources with their completion percentages in the corresponding menu.

Selecting a resource in the menu hot-reloads the translate view. Before loading we check for any unsaved changes in the editor, which we now also do upon clicking on the resource path in the Metadata section.

--

I still need to implement search and write some tests, and then I believe navigation will be in a good enough state for Stage 2 testing. We'll still need to figure out what to do with "All Projects" afterwards, because right now that view hides the resource menu and makes it hard to switch to the regular locale - project - resource view.

@adngdb I'd appreciate your help with the "Can't load specified string" message that appears when switching resources. For some reason `connected-react-router`'s `push()` method preserves the `entity` param and I haven't yet figured out how to reset it.

Also, a quick high-level review would be nice. :)

adngdb commented on frontend/src/modules/entitydetails/components/EntityDetails.js:
```diff
@@ -187,6 +188,17 @@ export class EntityDetailsBase extends React.Component<InternalProps, State> {
+                this.props.unsavedchanges,
+                () => { dispatch(push(path)); }
+            )
+        );
```
[rant] This pattern is repeated over a lot of places, and it makes me sad. I wished there was a way to refactor that so it's simpler to use (something like a Python decorator) but I don't know how that can be done. [/rant]

mathjazz replied:
Yeah. :/ Or a component lifecycle method.

adngdb commented on frontend/src/modules/entitydetails/components/Metadata.js:
```diff
@@ -217,9 +225,12 @@ export default class Metadata extends React.Component<Props> {
             <Localized id='entitydetails-metadata-resource' attrs={ { title: true } }>
                 <Property title='Resource' className='resource'>
-                    <Link to={ `/${locale.code}/${entity.project.slug}/${entity.path}/` }>
+                    <a
```
Why replace `Link` with `a`? If it's to use `onClick`, I'd be surprised if it didn't work with `Link`.

mathjazz replied:
What's the benefit of using `Link` instead of `a`? I'm preventing default anyways (to check for unsaved changes).

adngdb replied:
To me, it helps understanding what the link is about. I'm using `<a>` when I link to an external resource (a different page, a different website) and `<Link>` when I link to an internal resource, meaning when it doesn't trigger a page reload. It doesn't make much difference but I think it makes the intent clearer.

mathjazz replied:
Good point about semantics. I'll update.

mathjazz replied:
Meh, the problem with `<Link>` is that it prepends `/translate` to the `href` attribute, and it doesn't allow me to navigate to the right URL without manually stripping `/translate`.

adngdb commented on frontend/src/core/resource/components/ResourceMenu.js:
```diff
@@ -0,0 +1,164 @@
+        return Math.floor(resource.approved_strings / resource.total_strings * 100) + '%';
+    }
+
+    renderResource = (resource: Resource, i: number) => {
```
I've actually come to think that if you create a `renderX` method, you actually want to have an `X` component. In this case, it would be perfect to use a simple function component (instead of a class), because the code is so small. I'm more and more thinking that we should drop the rule of using only class-based components and embrace function components and hooks. What do you think?

mathjazz replied:
Yeah I was thinking about a component and I thought it was an overkill, but that's before I've been introduced to the concept of functional components. :) I'll give it a try.

mathjazz commented:
@adngdb I've addressed your comments, except the ones I commented on. Could you please have another look?

I've also:
* added the ability to search resources
* made percentages in the Resource menu update as you change translations
  * fixed two related bugs
* wrote some tests
* filed a bug for the ["Can't load specified string"](https://bugzilla.mozilla.org/show_bug.cgi?id=1553618) message (it's unrelated to this patch).

adngdb commented on frontend/src/core/resource/index.js:
```diff
@@ -0,0 +1,16 @@
+
+export { default as ResourceItem } from './components/ResourceItem';
+export { default as ResourceMenu } from './components/ResourceMenu';
+export { default as ResourcePercent } from './components/ResourcePercent';
```
`ResourceItem` and `ResourcePercent` should not be exposed, they are not meant to be used outside of the `resource` module.

adngdb commented on frontend/src/core/resource/components/ResourceMenu.js:
```diff
@@ -0,0 +1,172 @@
+
+import './ResourceMenu.css';
+
+import { ResourceItem, ResourcePercent } from '..';
```
You should import from the files directly, and not expose these outside this module.

adngdb commented on frontend/src/core/resource/components/ResourceMenu.js:
```diff
@@ -0,0 +1,172 @@
+    navigateToPath = (event: SyntheticMouseEvent<>) => {
+        event.preventDefault();
+
+        // $FLOW_IGNORE
```
Can you explain why you ignore Flow here?

mathjazz replied:
I don't think it's possible to circumvent the "Property pathname is missing in EventTarget [1]." Flow error. Is it?

adngdb replied:
Yeah, I guess so. But can you add a note to say that? :angel:

mathjazz replied:
Oh! Well, I can actually get rid of it. :)

adngdb commented on frontend/src/core/resource/components/ResourceMenu.js:
```diff
@@ -0,0 +1,172 @@
+            className += ' closed';
+        }
+
+        let resourceName = parameters.resource.split('/').slice(-1)[0];
```
I'm not sure I understand this. If there happens to be a resource `path/to/all-resources` then it will match the next condition, but I don't think it should. Why not just check if `parameters.resource` is equal to `all-resources`?

adngdb commented on frontend/src/core/resource/components/ResourceMenu.js:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        // Extract All Resources entry for more exposed presentation
+        const allResources = resources.resources.slice(-1)[0];
```
I assume the API returns a list of resources that also contains `all-resources`? If that's the case, I'd remove it in the reducer instead of here.

adngdb commented on frontend/src/core/resource/components/ResourceMenu.js:
```diff
@@ -0,0 +1,172 @@
+        );
+
+        return <li>
+            <div className={ className }>
```
Is this `<div>` needed? Can you move the class to the `<li>` instead and move everything one level down?

adngdb commented on frontend/src/core/resource/components/ResourceMenu.test.js:
```diff
@@ -0,0 +1,87 @@
+        expect(wrapper.find('.resource-menu .menu #navigation-ResourceMenu-all-projects')).toHaveLength(1);
+    });
+
+    it('renders resource menu correctly', () => {
```
You have the same test name twice here. This one should talk about search.

adngdb commented on frontend/src/core/resource/components/ResourcePercent.js:
```diff
@@ -0,0 +1,25 @@
+ * Render a resource item percentage.
+ */
+export default function ResourcePercent(props: Props) {
+    const { approved_strings, strings_with_warnings, total_strings }  = props.resource;
```
Double space before the `=` sign.

adngdb commented on frontend/src/core/resource/reducer.js:
```diff
@@ -0,0 +1,69 @@
+
+            const diff_approved = approved_strings - item.approved_strings;
+            item.approved_strings += diff_approved;
+            allResources.approved_strings += diff_approved;
```
`item` and `allResources` should be immutable. You should instead copy these objects first, and recreate a new list of resources based on them. Note that you don't need to copy the ones you do not modify.

adngdb replied:
I am very disappointed that Flow didn't catch that. :-(

adngdb commented on frontend/src/modules/entitydetails/components/Metadata.js:
```diff
@@ -202,6 +202,14 @@ export default class Metadata extends React.Component<Props> {
+    navigateToPath = (event: SyntheticMouseEvent<>) => {
+        event.preventDefault();
+
+        // $FLOW_IGNORE
```
Please explain why you ignore Flow here.

mathjazz replied:
See above.

adngdb commented on frontend/src/core/resource/components/ResourcePercent.js:
```diff
@@ -0,0 +1,25 @@
+ * Render a resource item percentage.
+ */
+export default function ResourcePercent(props: Props) {
+    const { approved_strings, strings_with_warnings, total_strings }  = props.resource;
```
Still a double-space before the equal sign. :-)

adngdb replied:
Forget that, I'm dumb.
